compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -71,10 +68,154 @@ getSwizzle(IREE::GPU::MMAIntrinsic mma, int operandIdx) {
                                               /*permutation=*/{1, 0, 2}};
     }
   default:
-    return std::nullopt;
+    assert(false && "should not get here.");
+    return {};
   }
 }
 
+// Given an `expandShape` vector-of-vectors describing the mapping from source
+// dimensions to expanded dimensions, returns the index of the first expanded
+// dimension corresponding to the given source dimension index.
+static int64_t
+getExpandedDimFirstIdx(const SmallVector<SmallVector<int64_t>> &expandShape,
+                       int64_t srcIndex) {
+  int dstIndexFirst = 0;
+  for (int i = 0; i < srcIndex; ++i) {
+    dstIndexFirst += expandShape[i].size();
+  }
+  return dstIndexFirst;
+}
+
+// Unroll the dimension given by `srcIndex` by the given `unrollFactor`.
+// This is not interleaving layouts. The layout will consist of multiple copies
+// of the input tile, side by side.
+//
+// Example:
+//    Input swizzle = { expandShape = [[16], [4]], permutation = [1, 0] }
+//    Input srcIndex = 1
+//    Input unrollFactor = 4
+// -> Output swizzle = { expandShape = [[16], [4, 4]], permutation = [1, 2, 0] }
+//
+static void unroll(MaterializeEncodingInfo::Swizzle &swizzle, int srcIndex,
+                   int unrollFactor) {
+  assert(unrollFactor > 1);
+  int dstIndexFirst = getExpandedDimFirstIdx(swizzle.expandShape, srcIndex);
+
+  // The new unrolling dimension is inserted at the start of the expandShape
+  // dimensions group corresponding to srcIndex.
+  swizzle.expandShape[srcIndex].insert(swizzle.expandShape[srcIndex].begin(),
+                                       unrollFactor);
+  // Since we are not interleaving here, generating side-by-side copies of the
+  // original layout, the new unrolling dimension is the new outermost
+  // dimension. Existing entries get shifted to make room for it.
+  for (auto &p : swizzle.permutation) {
+    p += (p >= dstIndexFirst);
+  }
+  swizzle.permutation.insert(swizzle.permutation.begin(), dstIndexFirst);
+}
+
+// Returns the index of the dimension whose flattened size (flattening inner
+// dimensions into it) matches the given `targetSize`.
+static int64_t getDimIdxForTargetSize(const SmallVector<int64_t> &shape,
+                                      int64_t targetSize) {
```

Comment:
This was a little hard for me to follow what this was for at first, but I think I understand it now. Maybe it would make it a bit easier to understand if you added an example. Some explanation of why it is needed (i.e., some intrinsics have a >1 number of elements for the thread to load contiguously, so there are extra inner dimensions) may also help.

Reply:
Added an example. But regarding the motivation, notice that this helper is generic, it is not in itself related to interleaving, so I didn't want the comment to lose sight of that. Instead, see response to the next comment, the call to this helper moves to the top-level `getSwizzle` so it inherently gets more context about how it's used.